### PR TITLE
Adding concept description of Endpoint and Endpoint Instance

### PR DESCRIPTION
**Pull request by WojcikMike**
PR created so that @mikeminutillo can help :stuck_out_tongue: . I made you an contributor.

Connects to #1237

Initiated by comment in to following PR: #1278

### CODE REVIEW

**Comment by WojcikMike** [comment]
I believe that Input Queue as concept that is related to Endpoint Instance should also live within endpoint category rather than messaging, which mean:
* moving this: http://docs.particular.net/nservicebus/messaging/specify-input-queue-name from messaging to endpoint
* Making hyperlink from Input Queue concept description to the above site:
http://docs.particular.net/nservicebus/concept-overview#input-queue

**Comment by mikeminutillo** [comment]
> moving this: http://docs.particular.net/nservicebus/messaging/specify-input-queue-name from messaging to endpoint

This requires a redirects header https://github.com/Particular/docs.particular.net#redirects

I agree that once we have a doco point for Endpoint there is a lot we can put in here (endpoint creation and startup, full vs. Send-Only, etc.)

File: nservicebus/endpoint/index.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+---
+title: Endpoint
+summary: Describe general concept of Endpoint and Endpoint Instance.
+tags:
+- Endpoint
+- Endpoint Instance
+---
+
```
Comment (mikeminutillo) [inline nservicebus/endpoint/index.md]: Should we have an intro para? Something really simple like

> A solution built on NServiceBus consists of a number of logical Endpoints that communicate via [Messaging](/nservicebus/messaging/).

**Comment by WojcikMike** [comment]
I am not sure what to write more in Endpoint/Endpoint Instance description. I referenced Configuring endpoint which shows differences in code.

**Comment by WojcikMike** [comment]
@Particular/docs-maintainers please review and suggest comments/improvements or merge. I addressed issues raised by @mikeminutillo

File: nservicebus/endpoint/index.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+## Endpoint
+
+An Endpoint is a design-time concept that has a name and a collection of associated Message Handlers and Sagas. A definition of an Endpoint lives as a group of artifacts in a source code repository. An Endpoint (the binaries that result from building the endpoint's source code) can be deployed to a number of machines and environments. Each such deployment is a separate Endpoint Instance.
+
+## Endpoint Instance
+
+An Endpoint Instance is a run-time object that allows to interact with the bus. Endpoint Instances are able to send and receive messages. It runs associated Message Handlers and Sagas to process incoming messages. An Endpoint Instance has a single [Input Queue](/nservicebus/endpoint/specify-input-queue-name.md) (which can be shared with other instances of the same Endpoint in a [scale-out scenario](/nservicebus/scalability-and-ha/scale-out.md)).
+
+Please check our samples to see how to [configure endpoint](/samples/endpoint-configuration/).
```
Comment (SimonCropp) [inline nservicebus/endpoint/index.md]: make this a related in the header

File: nservicebus/endpoint/index.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+---
+title: Endpoint
+summary: Describe general concept of Endpoint and Endpoint Instance.
```
Comment (SimonCropp) [inline nservicebus/endpoint/index.md]: i think general  is redundant

Comment (SimonCropp) [inline nservicebus/endpoint/index.md]: Describe the concepts of Endpoint and Endpoint Instance

File: nservicebus/endpoint/index.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+A solution built on NServiceBus consists of a number of logical Endpoints that communicate via [Messaging](/nservicebus/messaging).
+
+## Endpoint
+
+An Endpoint is a design-time concept that has a name and a collection of associated Message Handlers and Sagas. A definition of an Endpoint lives as a group of artifacts in a source code repository. An Endpoint (the binaries that result from building the endpoint's source code) can be deployed to a number of machines and environments. Each such deployment is a separate Endpoint Instance.
+
+## Endpoint Instance
+
+An Endpoint Instance is a run-time object that allows to interact with the bus. Endpoint Instances are able to send and receive messages. It runs associated Message Handlers and Sagas to process incoming messages. An Endpoint Instance has a single [Input Queue](/nservicebus/endpoint/specify-input-queue-name.md) (which can be shared with other instances of the same Endpoint in a [scale-out scenario](/nservicebus/scalability-and-ha/scale-out.md)).
```
Comment (SimonCropp) [inline nservicebus/endpoint/index.md]: perhaps

> An Endpoint Instance is a run-time object that allows interaction with the bus

Comment (SimonCropp) [inline nservicebus/endpoint/index.md]: perhaps

> Endpoint Instances are able to send, receive, and publish messages.

Comment (SimonCropp) [inline nservicebus/endpoint/index.md]: brackets not needed

> An Endpoint Instance has a single [Input Queue](/nservicebus/endpoint/specify-input-queue-name.md) which can be shared with other instances of the same Endpoint in a [scale-out scenario](/nservicebus/scalability-and-ha/scale-out.md.

File: nservicebus/endpoint/specify-input-queue-name.md [hunk 5]
```diff
@@ -6,6 +6,7 @@ tags:
 - Conventions
 redirects:
 - nservicebus/how-to-specify-your-input-queue-name
+- nservicebus/messaging/specify-input-queue-name
```
Comment (SimonCropp) [inline nservicebus/endpoint/specify-input-queue-name.md]: i think this rename will also require a link fix in `\nservicebus\ravendb\index.md`

File: nservicebus/endpoint/index.md (hunk 1 above)
Comment (SimonCropp) [inline nservicebus/endpoint/index.md]: @mikeminutillo  u could make the intro an include and re-use the text from concept-overview.md

File: nservicebus/endpoint/index.md (hunk 2 above)
Comment (mikeminutillo) [inline nservicebus/endpoint/index.md]: > make this a related in the header

And get rid of this line, right?

Comment (SimonCropp) [inline nservicebus/endpoint/index.md]: correct

File: nservicebus/endpoint/index.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+tags:
+- Endpoint
+- Endpoint Instance
+---
+
+A solution built on NServiceBus consists of a number of logical Endpoints that communicate via [Messaging](/nservicebus/messaging).
+
+## Endpoint
+
+An Endpoint is a design-time concept that has a name and a collection of associated Message Handlers and Sagas. A definition of an Endpoint lives as a group of artifacts in a source code repository. An Endpoint (the binaries that result from building the endpoint's source code) can be deployed to a number of machines and environments. Each such deployment is a separate Endpoint Instance.
```
Comment (mikeminutillo) [inline nservicebus/endpoint/index.md]: Is the part about artifacts in a source code repository adding any value? The fact that we then talk about an Endpoint as a compiled binary thing just adds to the confusion too. I wonder if we can't collapse both sections into something simpler.

>An Endpoint is a logical entity that communicates with other Endpoints via messaging. Each Endpoint has an identifying name and contains a collection of Message Handlers and Sagas. An Endpoint can be deployed to a number of machines and environments. Each such deployment is a separate Endpoint Instance. Each Endpoint Instance has an [Input Queue](/nservicebus/endpoint/specify-input-queue-name.md) which contains messages for the Endpoint Instance to process. 
>
>It is common for each Endpoint to have a single Endpoint Instance. As Endpoints need to [scale-out](/nservicebus/scalability-and-ha/scale-out.md), additional Endpoint Instances can be added. This collection of Endpoint Instances still represents a single logical Endpoint. 
>
>NOTE: In a scale-out scenario multiple Endpoint Instances can share a single Input Queue. 

Thoughts?

Comment (weralabaj) [inline nservicebus/endpoint/index.md]: I like your proposition. +1

File: nservicebus/endpoint/specify-input-queue-name.md (hunk 5 above)
Comment (WojcikMike) [inline nservicebus/endpoint/specify-input-queue-name.md]: Done

**Comment by WojcikMike** [comment]
I made all the changes, please review and merge in if it is in acceptable form @Particular/docs-maintainers

File: nservicebus/concept-overview.md [hunk 7]
```diff
@@ -30,14 +30,14 @@ The payload of the message is also called the body. It travels between the endpo
...
+An Endpoint is a logical entity that communicates with other Endpoints via messaging. Each Endpoint has an identifying name and contains a collection of Message Handlers and Sagas. An Endpoint can be deployed to a number of machines and environments. Each such deployment is a separate Endpoint Instance.
 
 
-### Endpoint Instance
+### [Endpoint Instance](/nservicebus/endpoint/)
 
-An Endpoint Instance is a run-time object that allows to interact with the bus. Endpoint Instances are able to send and receive messages. It runs associated Message Handlers and Sagas to process incoming messages. An Endpoint Instance has a single Input Queue (which can be shared with other instances of the same Endpoint in a scale-out scenario).
+An Endpoint Instance is a run-time object that allows interaction with the bus. Endpoint Instances are able to send, receive, and publish messages.. It runs associated Message Handlers and Sagas to process incoming messages. An Endpoint Instance has a single Input Queue which can be shared with other instances of the same Endpoint in a scale-out scenario.
```
Comment (indualagarsamy) [inline nservicebus/concept-overview.md]: Minor corrections. - Extra . Missing , 

Although I have no idea what this means:
> It runs associated Message Handlers and Sagas to process incoming messages. 

My suggestion is below: 

An Endpoint Instance is a run-time object that allows interaction with the bus. Endpoint Instances can send, receive, and publish messages. It runs associated Message Handlers and Sagas to process incoming messages. An Endpoint Instance has a single Input Queue, which can be shared with other instances of the same Endpoint in a scale-out scenario.

File: nservicebus/endpoint/index.md [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+title: Endpoint
+summary: Describe the concepts of Endpoint and Endpoint Instance.
+tags:
+- Endpoint
+- Endpoint Instance
+related:
+ - samples/endpoint-configuration
+---
+
+An Endpoint is a logical entity that communicates with other Endpoints via [Messaging](/nservicebus/messaging). Each Endpoint has an identifying name and contains a collection of Message Handlers and Sagas. An Endpoint can be deployed to a number of machines and environments. Each such deployment is a separate Endpoint Instance. Each Endpoint Instance has an [Input Queue](/nservicebus/endpoint/specify-input-queue-name.md) which contains messages for the Endpoint Instance to process.
```
Comment (indualagarsamy) [inline nservicebus/endpoint/index.md]: I think we should link here to both Message Handlers and Sagas as well.

File: nservicebus/concept-overview.md (hunk 7 above)
Comment (indualagarsamy) [inline nservicebus/concept-overview.md]: > An Endpoint Instance has a single Input Queue which can be shared with other instances of the same Endpoint in a scale-out scenario.

Is this true for MSMQ? True, an endpoint instance has a single input queue, but is it shared with other instances?

File: nservicebus/concept-overview.md [hunk 9]
```diff
@@ -30,14 +30,14 @@ The payload of the message is also called the body. It travels between the endpo
 Additional information about a message is communicated over the transport in a collection of key value pairs. Message headers are similar to HTTP headers in the sense that they define how messages should be transmitted and processed. NServiceBus uses headers to implement some of its features. Users are also able to add [custom headers](/nservicebus/messaging/header-manipulation.md).
 
 
-### Endpoint
+### [Endpoint](/nservicebus/endpoint/)
 
-An Endpoint is a design-time concept that has a name and a collection of associated Message Handlers and Sagas. A definition of an Endpoint lives as a group of artifacts in a source code repository. An Endpoint (the binaries that result from building the endpoint's source code) can be deployed to a number of machines and environments. Each such deployment is a separate Endpoint Instance.
+An Endpoint is a logical entity that communicates with other Endpoints via messaging. Each Endpoint has an identifying name and contains a collection of Message Handlers and Sagas. An Endpoint can be deployed to a number of machines and environments. Each such deployment is a separate Endpoint Instance.
```
Comment (indualagarsamy) [inline nservicebus/concept-overview.md]: `number of machines` can be `several` machines.

>Each such deployment is a separate Endpoint Instance.
doesn't read well.

How about:

An Endpoint is a logical entity that communicates with other Endpoints via messaging. Each Endpoint has an identifying name and contains a collection of Message Handlers and Sagas. An Endpoint can be deployed to several machines and environments. Each deployment of an endpoint is an instance.

File: nservicebus/endpoint/index.md [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+- Endpoint Instance
+related:
+ - samples/endpoint-configuration
+---
+
+An Endpoint is a logical entity that communicates with other Endpoints via [Messaging](/nservicebus/messaging). Each Endpoint has an identifying name and contains a collection of Message Handlers and Sagas. An Endpoint can be deployed to a number of machines and environments. Each such deployment is a separate Endpoint Instance. Each Endpoint Instance has an [Input Queue](/nservicebus/endpoint/specify-input-queue-name.md) which contains messages for the Endpoint Instance to process. 
+
+It is common for each Endpoint to have a single Endpoint Instance. As Endpoints need to [scale-out](/nservicebus/scalability-and-ha/scale-out.md), additional Endpoint Instances can be added. This collection of Endpoint Instances still represents a single logical Endpoint. 
+
+NOTE: In a scale-out scenario multiple Endpoint Instances can share a single Input Queue.
```
Comment (indualagarsamy) [inline nservicebus/endpoint/index.md]: > NOTE: In a scale-out scenario multiple Endpoint Instances can share a single Input Queue. 

Isn't this specific to non-MSMQ transports?

**Comment by indualagarsamy** [comment]
@WojcikMike - If you create this either as a local branch or grant me access to your fork, I can make some of my recommended changes and try and get this merged.

File: nservicebus/endpoint/index.md (hunk 10 above)
Comment (WojcikMike) [inline nservicebus/endpoint/index.md]: As above, it isn't. And the sentence only tells that "can share". I honestly believe that this should be described in some other place. I left this sentence here.

**Comment by WojcikMike** [comment]
I made suggested changes, I added contributor rights to all maintainers to my fork. 

We are pending with the following:
* What should we do with: 

> "NOTE: In a scale-out scenario multiple Endpoint Instances can share a single Input Queue." 

I honestly believe that it should be covered somewhere in scale-out articles rather than in here. But should we leave it as is, and create separate issue to work on that doco? Or should we remove this note entirely?

**Comment by weralabaj** [comment]
I think it should be removed from here. The info about queues could be moved to articles dedicated to particular transports, e.g. for SQL Server transport we have this http://docs.particular.net/nservicebus/sqlserver/design#queues. It doesn't explicitly say it's talking about queues in scale-out, but I think that level of detail is enough for most users.

So let's open a new issues for other transports to clarify?

**Comment by weralabaj** [comment]
I've raised the issue here https://github.com/Particular/docs.particular.net/issues/1317. Closing this one.

**Pull request merged**